README.rst
Django Freckle Budgets
======================

A reusable Django app that allows to plan year budgets and track them via letsfreckle.

Installation
------------

To get the latest stable release from PyPi

.. code-block:: bash

    pip install django-freckle-budgets

To get the latest commit from GitHub

.. code-block:: bash

    pip install -e git+git://github.com/bitmazk/django-freckle-budgets.git#egg=freckle_budgets

TODO: Describe further installation steps (edit / remove the examples below):

Add ``freckle_budgets`` to your ``INSTALLED_APPS``

.. code-block:: python

    INSTALLED_APPS = (
        ...,
        'freckle_budgets',
    )

Add the ``freckle_budgets`` URLs to your ``urls.py``

.. code-block:: python

    urlpatterns = patterns('',
        ...
        url(r'^freckle-budgets/', include('freckle_budgets.urls')),
    )

Don't forget to migrate your database

.. code-block:: bash

    ./manage.py migrate freckle_budgets

Finally make sure to add the necessary settings to your ``local_settings.py``:


.. code-block:: python
    FRECKLE_BUDGETS_ACCOUNT_NAME = 'your-freckle-account-name' 
    FRECKLE_BUDGETS_API_TOKEN = 'your-freckle-api-token' 


Usage
-----

Add a year
++++++++++

The first thing you need to do is to add a year via the Django admin:

* **Year**: This is the year you want to add, i.e. 2015
* **Rate**: This is your hourly rate that you would like to get for new
  projects in this year. This value will be used to calculate the potential
  amount of money you could earn if you were fully booked out (one can be
  allowed to dream).
* **Work hours per day**: The number of hours you will definitely work on each
  work day.
* **Sick leave days**: The number of sick leave days that you estimate (this is
  per person, not the total for all employees) 
* **Vacation days**: The number of vacation days that each employee may take
  during this year.

These values will be used to calculate the total number of hours that your team
can possibly work during any month.

Add months
++++++++++

Once you have added a year, you need to add twelve months for that year via the
Django admin:

* **Year**: The year this month belongs to.
* **Month**: The month number (i.e. 1 for January)
* **Number of employees**: The number of employees that you will have during
  this month.
* **Public holidays**: The number of public holidays that this month has in
  this year.

Again, these values are mainly needed to calculate the total number of hours
that your team can work during this month.

Add projects
++++++++++++

After adding a year and it's months, you need to add a few projects via the
Django admin:

* **Name**: The display name for this project.
* **Freckle project ID**: The project ID of this project in your freckle
  account.
* **Color**: A CSS color code for this project (i.e. ``#C7B06F``)
* **Is investment**: If you are working on this project for free (i.e. for
  internal projects), enable this checkbox.

Add project budgets to months
+++++++++++++++++++++++++++++

For each of your projects, add as many ``ProjectMonth`` objects as needed. For
example, if you have a long running project that has budgets for the whole
year, you would have to add twelve objects:

* **Project**: The project you want to plan for.
* **Month**: The month you want to plan for.
* **Budget**: The budget that you can burn on this project during this month.
* **Rate**: The hourly rate that you can bill for this project.

Contribute
----------

If you want to contribute to this project, please perform the following steps

.. code-block:: bash

    # Fork this repository
    # Clone your fork
    mkvirtualenv -p python2.7 django-freckle-budgets
    make develop

    git co -b feature_branch master
    # Implement your feature and tests
    git add . && git commit
    git push -u origin feature_branch
    # Send us a pull request for your feature branch

In order to run the tests, simply execute ``tox``. It will create two more
venvs and run tests against Django 1.6 and Django 1.7.


CHANGELOG.txt
=== 0.0.X (onggoing, to be released as 0.2) ===

- Updated admins
- Properly counting times for investment projects

=== 0.0.X (onggoing, to be released as 0.1) ===
- Initial commit


# Suggested file syntax:
#
# === (ongoing) ===
# - this is always on top of the file
# - when you release a new version, you rename the last `(ongoing)` to the new
#   version and add a new `=== (ongoing) ===` to the top of the file
#
# === 1.0 ===
# - a major version is created when the software reached a milestone and is 
#   feature complete
#
# === 0.2 ===
# - a minor version is created when a lot of new features have bene added or
#   significant backwards incompatible changes have been made.


freckle_budgets/tests/fixtures.py
"""Test-fixtures for the freckle_budgets app."""
from mixer.backend.django import mixer


def create_month(cls):
    """Creates one ``Month`` object."""
    cls.year = mixer.blend(
        'freckle_budgets.Year', year=2015, sick_leave_days=12,
        vacation_days=12, rate=100)
    cls.month = mixer.blend(
        'freckle_budgets.Month', month=1, year=cls.year, public_holidays=1,
        employees=2)


def create_project_months(cls):
    """Creates two ``ProjectMonth`` objects."""
    create_month(cls)
    cls.month2 = mixer.blend(
        'freckle_budgets.Month', month=2, year=cls.year, public_holidays=1,
        employees=2)
    cls.proj1 = mixer.blend(
        'freckle_budgets.Project', is_investment=False, freckle_project_id='1')
    cls.proj2 = mixer.blend(
        'freckle_budgets.Project', is_investment=True, freckle_project_id='2')

    cls.project_month1_1 = mixer.blend(
        'freckle_budgets.ProjectMonth', project=cls.proj1,
        month=cls.month, budget=1000, rate=100, )
    cls.project_month2_1 = mixer.blend(
        'freckle_budgets.ProjectMonth', project=cls.proj1,
        month=cls.month2, budget=2000, rate=100, )

    cls.project_month1_2 = mixer.blend(
        'freckle_budgets.ProjectMonth', project=cls.proj2,
        month=cls.month, budget=4000, rate=200, )


def create_employee_project_months(cls):
    """Creates fixtures that allows to calculate an employees workload."""
    create_project_months(cls)
    cls.employee1 = mixer.blend('freckle_budgets.Employee')
    cls.employee_project_month = mixer.blend(
        'freckle_budgets.EmployeeProjectMonth',
        project_month=cls.project_month1_1, employee=cls.employee1,
        responsibility=100)

    cls.employee2 = mixer.blend('freckle_budgets.Employee')
    cls.employee_project_month = mixer.blend(
        'freckle_budgets.EmployeeProjectMonth',
        project_month=cls.project_month1_2, employee=cls.employee2,
        responsibility=50)


def get_api_response(cls):
    """
    Returns a typical freckle API response and creates necessary objects.

    """
    cls.year = mixer.blend('freckle_budgets.Year', year=2015)
    cls.month1 = mixer.blend('freckle_budgets.Month', month=1, year=cls.year)
    cls.month2 = mixer.blend('freckle_budgets.Month', month=2, year=cls.year)
    cls.proj1 = mixer.blend(
        'freckle_budgets.Project', freckle_project_id='111')
    cls.proj2 = mixer.blend(
        'freckle_budgets.Project', freckle_project_id='222')
    cls.proj3 = mixer.blend(
        'freckle_budgets.Project', freckle_project_id='333',
        is_investment=True)

    cls.proj1_month1 = mixer.blend(
        'freckle_budgets.ProjectMonth', project=cls.proj1, month=cls.month1,
        budget=1000, rate=100)
    cls.proj1_month2 = mixer.blend(
        'freckle_budgets.ProjectMonth', project=cls.proj1, month=cls.month2,
        budget=1000, rate=100)
    cls.proj2_month1 = mixer.blend(
        'freckle_budgets.projectmonth', project=cls.proj2, month=cls.month1,
        budget=2000, rate=200)
    cls.proj2_month2 = mixer.blend(
        'freckle_budgets.projectmonth', project=cls.proj2, month=cls.month2,
        budget=2000, rate=200)
    cls.proj3_month1 = mixer.blend(
        'freckle_budgets.ProjectMonth', project=cls.proj3, month=cls.month1,
        budget=4000, rate=400)
    cls.proj3_month2 = mixer.blend(
        'freckle_budgets.ProjectMonth', project=cls.proj3, month=cls.month2,
        budget=4000, rate=400)

    entries = [
        {
            # First project, first month, billable hours
            'entry': {
                'date': '2015-01-01',
                'project_id': 111,
                'billable': True,
                'minutes': 1,
            }
        },
        {
            # Unbillable hours should not be added up
            'entry': {
                'date': '2015-01-01',
                'project_id': 111,
                'billable': False,
                'minutes': 2,
            }
        },
        {
            # Billable hours should be added up
            'entry': {
                'date': '2015-01-01',
                'project_id': 111,
                'billable': True,
                'minutes': 4,
            }
        },
        {
            # Another project in the same month
            'entry': {
                'date': '2015-01-01',
                'project_id': 222,
                'billable': True,
                'minutes': 8,
            }
        },
        {
            # Another month
            'entry': {
                'date': '2015-02-01',
                'project_id': 222,
                'billable': True,
                'minutes': 16,
            }
        },
        {
            # Unbillable hours for investment projects should be added up
            'entry': {
                'date': '2015-02-01',
                'project_id': 333,
                'billable': False,
                'minutes': 32,
            }
        },
    ]
    return entries


freckle_budgets/tests/models_tests.py
"""Tests for the models of the freckle_budgets app."""
import datetime

from django.test import TestCase

from mixer.backend.django import mixer

from . import fixtures
from .. import models


class EmployeeTestCase(TestCase):
    """Tests for the ``Employee`` model."""
    longMessage = True

    def test_class(self):
        obj = mixer.blend('freckle_budgets.Employee')
        self.assertTrue(obj.pk, msg=('Object can be saved'))
        self.assertEqual(obj.__str__(), str(obj.name), msg=(
            '__str__ should return correct string'))


class YearTestCase(TestCase):
    """Tests for the ``Year`` model."""
    longMessage = True

    def test_model(self):
        obj = mixer.blend('freckle_budgets.Year')
        self.assertTrue(obj.pk, msg=('Object can be saved'))
        self.assertEqual(obj.__str__(), str(obj.year), msg=(
            '__str__ should return correct string'))

    def test_get_total_cashflow_budget(self):
        fixtures.create_project_months(self)
        result = self.year.get_total_cashflow_budget()
        self.assertEqual(result, 3000, msg=(
            'Should add up all planned budgets for all projects that have'
            ' is_investment==False for this year'))

    def test_get_total_investment_budget(self):
        fixtures.create_project_months(self)
        result = self.year.get_total_investment_budget()
        self.assertEqual(result, 4000, msg=(
            'Should add up all planned budgets for all projects that have'
            ' is_investment==True for this year'))

    def test_get_total_unused_budget(self):
        fixtures.create_project_months(self)
        result = self.year.get_total_unused_budget()
        self.assertEqual(result, 31000, msg=(
            'Should add up all unused budgets for all months of the year'))


class MonthManagerTestCase(TestCase):
    """Tests for the ``MonthManager`` model manager."""
    longMessage = True

    def test_get_months_with_projects(self):
        fixtures.create_project_months(self)
        mixer.blend('freckle_budgets.Month', month=3, year=self.year)

        manager = models.Month.objects
        result = manager.get_months_with_projects(2015)
        self.assertEqual(result.count(), 2, msg=(
            'Should return all months that have a project'))


class MonthTestCase(TestCase):
    """Tests for the ``Month`` model."""
    longMessage = True

    def test_model(self):
        obj = mixer.blend('freckle_budgets.Month')
        self.assertTrue(obj.pk, msg=('Obj can be saved'))

        expected = '{0} {1}'.format(obj.month, obj.year)
        self.assertEqual(obj.__str__(), expected, msg=(
            '__str__ should return correct string'))

    def test_get_available_ressources(self):
        fixtures.create_month(self)
        result = self.month.get_available_ressources()
        self.assertEqual(result, 190, msg=(
            'January has 22 work days. Minus holidays, vacations and sick'
            ' leave, 19 days remain. Multiplied with the daily hours (5) and'
            ' number of employees (2), the result should be 19 * 5 * 2 = 190'))

    def test_get_available_ressources_max(self):
        fixtures.create_month(self)
        result = self.month.get_available_ressources_max()
        self.assertEqual(result, 210, msg=(
            'January has 22 work days. Minus holidays, 21 days remain.'
            ' Multiplied with the daily hours (5) and number of employees (2),'
            ' the result should be 21 * 5 * 2 = 210'))

    def test_get_available_ressources_max_per_person(self):
        fixtures.create_month(self)
        result = self.month.get_available_ressources_max_per_person()
        self.assertEqual(result, 105, msg=(
            'January has 22 work days. Minus holidays, 21 days remain.'
            ' Multiplied with the daily hours (5) the result should be'
            ' 21 * 5 = 105'))

    def test_get_average_rate(self):
        fixtures.create_project_months(self)
        result = self.month.get_average_rate()
        self.assertEqual(result, 100, msg=(
            'Should divide the total profit by the total cashflow hours'))

    def test_get_cashflow_projects(self):
        fixtures.create_project_months(self)
        result = self.month.get_cashflow_projects()
        self.assertEqual(result.count(), 1, msg=(
            'Should return all projects that have is_investment=False'))
        self.assertFalse(result[0].project.is_investment, msg=(
            'Should return all projects that have is_investment=False'))

    def test_get_date(self):
        fixtures.create_month(self)
        result = self.month.get_date()
        expected = datetime.date(2015, 1, 1)
        self.assertEqual(result, expected, msg=(
            'Should return a date object for this month'))

    def test_get_investment_projects(self):
        fixtures.create_project_months(self)
        result = self.month.get_investment_projects()
        self.assertEqual(result.count(), 1, msg=(
            'Should return all projects that have is_investment=True'))
        self.assertTrue(result[0].project.is_investment, msg=(
            'Should return all projects that have is_investment=True'))

    def test_get_total_cashflow_hours(self):
        fixtures.create_project_months(self)
        result = self.month.get_total_cashflow_hours()
        self.assertEqual(result, 10, msg=(
            'Should add up all budget hours for each cashflow project (the'
            ' hours that can be spend on projects for the given budget and'
            ' rate)'))

    def test_get_total_investment_hours(self):
        fixtures.create_project_months(self)
        result = self.month.get_total_investment_hours()
        self.assertEqual(result, 20, msg=(
            'Should add up all budget hours for each investment project (the'
            ' hours that can be spend on projects for the given budget and'
            ' rate)'))

    def test_get_total_investment(self):
        fixtures.create_project_months(self)
        result = self.month.get_total_investment()
        self.assertEqual(result, 4000, msg=(
            'Should add up all budgets for all investment projects.'))

    def test_get_total_profit(self):
        fixtures.create_project_months(self)
        result = self.month.get_total_profit()
        self.assertEqual(result, 1000, msg=(
            'Should add up all budgets for all cashflow projects.'))

    def test_get_unused_hours(self):
        fixtures.create_project_months(self)
        result = self.month.get_unused_hours()
        self.assertEqual(result, 160, msg=(
            'Should add up all budget hours for all projects in this month,'
            ' then substract this from the total available ressources of this'
            ' month.'))

    def test_get_unused_daily_hours(self):
        fixtures.create_project_months(self)
        result = self.month.get_unused_daily_hours()
        self.assertEqual(result, 160.0 / 19, msg=(
            'Should take the total unused hours and divide them by the number'
            ' of work days in this month'))

    def test_get_unused_budget(self):
        fixtures.create_project_months(self)
        result = self.month.get_unused_budget()
        self.assertEqual(result, 160.0 * 100, msg=(
            'Should multiply the unused hours with the default rate for this'
            ' year.'))

    def test_get_weekdays(self):
        fixtures.create_month(self)
        result = self.month.get_weekdays()
        self.assertEqual(result, 22, msg=(
            'Should return 22 weekdays for January 2015'))

    def test_get_work_days(self):
        fixtures.create_month(self)
        result = self.month.get_work_days()
        self.assertEqual(result, 19, msg=(
            'January has 22 work days. Minus one public holiday. Minus 1'
            ' vacation day (12 / 12). Minus one sick leave day (12 / 12).'
            ' Therefore, the result should be 22 - 1 - 1 -1 = 19'))

        result = self.month.get_work_days(minus_sick_leave=False)
        self.assertEqual(result, 20, msg=(
            'Like above, but this time we do not substract the sick-leave'
            ' days'))

        result = self.month.get_work_days(
            minus_sick_leave=False, minus_vacations=False)
        self.assertEqual(result, 21, msg=(
            'Like above, but this time we do not substract the sick-leave'
            ' and vacation days'))

    def test_get_work_days_max(self):
        fixtures.create_month(self)
        result = self.month.get_work_days_max()
        self.assertEqual(result, 21, msg=(
            'January has 22 work days. Minus public holidays, the result'
            ' should be 21'))

    def test_get_workload(self):
        fixtures.create_employee_project_months(self)
        result = self.month.get_workloads()
        expected = self.project_month1_1.get_daily_hours()
        self.assertEqual(
            result[self.employee1.freckle_id]['cashflow_workload'], expected,
            msg=('When the employee has 100% responsibility for the project,'
                 ' the result should equal the daily hours for that project'
                 ' in this month'))

        expected = self.project_month1_2.get_daily_hours() * 0.5
        self.assertEqual(
            result[self.employee2.freckle_id]['investment_workload'], expected,
            msg=('When the employee has 50% responsibility for the project,'
                 ' the result should equal 50% of the daily hours for that'
                 ' project in this month'))


class ProjectManagerTestCase(TestCase):
    """Tests for the ``ProjectManager`` model manager."""
    longMessage = True

    def test_get_for_year(self):
        fixtures.create_project_months(self)
        mixer.blend(
            'freckle_budgets.Project', freckle_project_id='101')
        manager = models.Project.objects
        result = manager.get_for_year(2015)
        self.assertEqual(result.count(), 2, msg=(
            'Should return all projects that have ProjectMonth instances for'
            ' the given year'))


class ProjectTestCase(TestCase):
    """Tests for the ``Project`` model."""
    longMessage = True

    def test_model(self):
        obj = mixer.blend('freckle_budgets.Project')
        self.assertTrue(obj.pk, msg=('Object can be saved'))
        self.assertEqual(obj.__str__(), obj.name, msg=(
            '__str__ should return correct string'))


class ProjectMonthManagerTestCase(TestCase):
    """Tests for the ``ProjectManager` model manager."""
    longMessage = True

    def test_get_project_months(self):
        fixtures.create_project_months(self)
        mixer.blend('freckle_budgets.Month', month=2, year__year=2015)
        manager = models.ProjectMonth.objects
        result = manager.get_project_months(2015)
        self.assertEqual(result[0][0], self.month, msg=(
            'Should return a list of lists. The inner lists consist of a month'
            ' and a list of projects for this month.'))

    def test_get_freckle_projects(self):
        fixtures.create_project_months(self)
        manager = models.ProjectMonth.objects
        result = manager.get_freckle_projects(2015)
        self.assertEqual(list(result), [u'1', u'2'], msg=(
            'Should return all freckle project IDs for this month'))


class ProjectMonthTestCase(TestCase):
    """Tests for the ``ProjectMonth`` model."""
    longMessage = True

    def test_model(self):
        obj = mixer.blend('freckle_budgets.ProjectMonth')
        self.assertTrue(obj.pk, msg=('Object can be saved'))

        result = obj.__str__()
        expected = '{0} - {1}'.format(obj.project, obj.month)
        self.assertEqual(result, expected, msg=(
            '__str__ should return correct string'))

    def test_get_budget_hours(self):
        fixtures.create_project_months(self)

        result = self.project_month1_1.get_budget_hours()
        expected = self.project_month1_1.budget / self.project_month1_1.rate
        self.assertEqual(result, expected, msg=(
            'Should return the hours that can be worked on this project given'
            ' the budget and hourly rate'))

        self.project_month1_1.overhead_previous_month = \
            self.project_month1_1.budget / 2
        self.project_month1_1.save()
        expected = expected / 2
        result = self.project_month1_1.get_budget_hours()
        self.assertEqual(result, expected, msg=(
            'Should return the hours that can be worked on this project given'
            ' the budget and hourly rate, should deduct eventual overhead'
            ' hours from the budget when calculating the result'))

    def test_get_daily_hours(self):
        fixtures.create_project_months(self)
        pm = self.project_month1_1
        result = pm.get_daily_hours()
        expected = pm.get_budget_hours() / pm.month.get_work_days()
        self.assertEqual(result, expected, msg=(
            'Should return the daily hours that need to be worked on this'
            ' given the project budget, rate and work days of this month.'))


class EmployeeProjectMonthTestCase(TestCase):
    """Tests for the ``EmployeeProjectMonth`` model."""
    longMessage = True

    def test_model(self):
        obj = mixer.blend('freckle_budgets.EmployeeProjectMonth')
        self.assertTrue(obj.pk, msg=('Object can be saved'))

        expected = '{0} - {1}'.format(obj.project_month, obj.employee)
        self.assertEqual(obj.__str__(), expected, msg=(
            '__str__ should return correct string'))


freckle_budgets/models.py
"""Models for the freckle_budgets app."""
import calendar
import datetime

from django.db import models
from django.db.models import Sum
from django.utils.translation import ugettext_lazy as _
from django.utils.encoding import python_2_unicode_compatible


@python_2_unicode_compatible
class Employee(models.Model):
    name = models.CharField(max_length=256, verbose_name=_('Name'))
    freckle_id = models.CharField(max_length=256, verbose_name=_('Freckle ID'))

    class Meta:
        ordering = ['name', ]

    def __str__(self):
        return str(self.name)


@python_2_unicode_compatible
class Year(models.Model):
    year = models.PositiveIntegerField(verbose_name=_('Year'))
    rate = models.FloatField(default=100, verbose_name=_('Rate'))
    work_hours_per_day = models.FloatField(
        default=5, verbose_name=_('Work hours per day'))
    sick_leave_days = models.PositiveIntegerField(
        verbose_name=_('Sick leave days'))
    vacation_days = models.PositiveIntegerField(
        verbose_name=_('Vacation days'))

    class Meta:
        ordering = ['-year', ]

    def __str__(self):
        return str(self.year)

    def get_total_cashflow_budget(self):
        return ProjectMonth.objects.filter(
            month__year=self, project__is_investment=False).aggregate(
                Sum('budget'))['budget__sum']

    def get_total_investment_budget(self):
        return ProjectMonth.objects.filter(
            month__year=self, project__is_investment=True).aggregate(
                Sum('budget'))['budget__sum']

    def get_total_unused_budget(self):
        result = 0
        for month in Month.objects.filter(year=self):
            result += month.get_unused_budget()
        return result


class MonthManager(models.Manager):
    def get_months_with_projects(self, year):
        qs = self.get_queryset()
        return qs.filter(year__year=year, projectmonth__isnull=False).order_by(
            'month').distinct()


@python_2_unicode_compatible
class Month(models.Model):
    year = models.ForeignKey(Year, verbose_name=_('Year'))
    month = models.PositiveIntegerField(verbose_name=_('Month'))
    employees = models.FloatField(
        default=1, verbose_name=_('Number of employees'))
    public_holidays = models.PositiveIntegerField(
        verbose_name=_('Public holydays'), default=0, null=True)

    objects = MonthManager()

    class Meta:
        ordering = ['-year', 'month']

    def __str__(self):
        return '{0} {1}'.format(self.month, self.year)

    def get_available_ressources(self):
        """Returns the available hours that the team can work this month."""
        work_days = self.get_work_days()
        work_hours = self.year.work_hours_per_day * work_days * self.employees
        return work_hours

    def get_available_ressources_max(self):
        """Returns the available hours that the team can work this month."""
        work_days = self.get_work_days(
            minus_sick_leave=False, minus_vacations=False)
        work_hours = self.year.work_hours_per_day * work_days * self.employees
        return work_hours

    def get_available_ressources_max_per_person(self):
        """Returns the hours that each employee should work this month."""
        return self.get_available_ressources_max() / self.employees * 1.0

    def get_average_rate(self):
        """Returns the average rate for all cashflow projects of this month."""
        return self.get_total_profit() / self.get_total_cashflow_hours()

    def get_cashflow_projects(self):
        return ProjectMonth.objects.filter(month=self).exclude(
            project__is_investment=True).order_by('-budget', )

    def get_date(self):
        return datetime.date(self.year.year, self.month, 1)

    def get_investment_projects(self):
        return ProjectMonth.objects.filter(
            month=self, project__is_investment=True).order_by('-budget', )

    def get_total_cashflow_hours(self):
        projects = self.get_cashflow_projects()
        hours = 0
        for project in projects:
            hours += project.get_budget_hours()
        return hours

    def get_total_investment_hours(self):
        projects = self.get_investment_projects()
        hours = 0
        for project in projects:
            hours += project.get_budget_hours()
        return hours

    def get_total_investment(self):
        projects = self.get_investment_projects()
        profit = 0
        for project in projects:
            profit += project.budget
        return profit

    def get_total_profit(self):
        projects = self.get_cashflow_projects()
        profit = 0
        for project in projects:
            profit += project.budget
        return profit

    def get_unused_hours(self):
        projects = ProjectMonth.objects.filter(month=self)
        used_hours = 0
        for project in projects:
            used_hours += project.get_budget_hours()
        unused_hours = self.get_available_ressources() - used_hours
        return unused_hours

    def get_unused_daily_hours(self):
        unused_hours = self.get_unused_hours()
        return unused_hours / self.get_work_days()

    def get_unused_budget(self):
        return self.get_unused_hours() * self.year.rate

    def get_weekdays(self):
        """Returns the number of weekdays of this month."""
        month_dates = calendar.Calendar(0).itermonthdates(
            self.year.year, self.month)
        weekdays = 0
        for day in month_dates:
            if day.month != self.month:
                continue
            if day.weekday() < 5:
                weekdays += 1
        return weekdays

    def get_work_days(self, minus_sick_leave=True, minus_vacations=True):
        """
        Returns the work days that should be used in whole-year-calculations.

        Public holidays, sick leave days and vacation days will be deducted.

        :minus_sick_leave: If ``False``, sick leave will not be deducted from
          the total work days.
        :minus_vacations: If ``False``, vacations will not be deducted from
          the total work days.

        """
        work_days = self.get_weekdays() * 1.0
        work_days -= self.public_holidays
        if minus_sick_leave:
            work_days -= self.year.sick_leave_days / 12
        if minus_vacations:
            work_days -= self.year.vacation_days / 12
        return work_days

    def get_work_days_max(self):
        """
        Returns the maximum work days, if no one falls sick or takes vacations.

        """
        return self.get_work_days(
            minus_sick_leave=False, minus_vacations=False)

    def get_workloads(self):
        """
        Returns the workload for all employees for this month.

        The workload is an amount of hours that the employee needs to work per
        available work day in order to meet all planned budgets.

        The workload is calculated as follows:

        * Project1: Budget = 1000
        * Project1: Rate = 100
        * Project1: Budget time = 10 hours
        * Employee responsibility for the project: 50%
        * Available work days for the month: 18
        * Employee work load = 10 / 18 * 0.5 = 0.28

        The result looks like this:

            {
                employee_freckle_id: {
                    'name': Heino,
                    'cashflow_workload': 4,
                    'investment_workload': 2,
                    'total_workload': 6,
                },
                employee_freckle_id: { ... },
                ...
            }

        """
        result = {}
        employee_project_months = EmployeeProjectMonth.objects.filter(
            project_month__month=self)
        for epm in employee_project_months:
            if epm.employee.freckle_id not in result:
                result[epm.employee.freckle_id] = {
                    'name': epm.employee.name,
                    'cashflow_workload': 0,
                    'investment_workload': 0,
                    'total_workload': 0,
                }

            daily_hours = epm.project_month.get_daily_hours()
            if epm.project_month.project.is_investment:
                key = 'investment_workload'
            else:
                key = 'cashflow_workload'

            workload = daily_hours * (epm.responsibility / 100.0)
            result[epm.employee.freckle_id][key] += workload
            result[epm.employee.freckle_id]['total_workload'] += workload
        return result


class ProjectManager(models.Manager):
    def get_for_year(self, year):
        qs = self.get_queryset()
        qs = qs.filter(
            project_months__month__year__year__exact=year).distinct()
        return qs


@python_2_unicode_compatible
class Project(models.Model):
    name = models.CharField(max_length=100, verbose_name=_('Name'))
    freckle_project_id = models.CharField(
        max_length=256, verbose_name=_('Freckle project ID'), null=True)
    color = models.CharField(max_length=100, verbose_name=_('Color'))
    is_investment = models.BooleanField(
        default=False, verbose_name=_('Is investment'))

    objects = ProjectManager()

    class Meta:
        ordering = ['name', ]

    def __str__(self):
        return self.name


class ProjectMonthManager(models.Manager):
    def get_project_months(self, year):
        result = []
        months = Month.objects.get_months_with_projects(year)
        for month in months:
            project_months = self.get_queryset().filter(month=month)
            result.append([month, project_months])
        return result

    def get_freckle_projects(self, year):
        qs = self.get_queryset()
        return qs.filter(month__year__year=year).order_by(
            'project__freckle_project_id').values_list(
                'project__freckle_project_id', flat=True).distinct()


@python_2_unicode_compatible
class ProjectMonth(models.Model):
    """
    Contains the budget for a certain project in a certain month.

    :project: A ``Project`` instance.
    :month: A ``Month`` instance.
    :budget: The budget for this porject in this month.
    :rate: The rate that can be billed for this project in this month. Budget
      divided by rate equals ``budget_hours``.
    :overhead_previous_month: Sometimes you have already worked hours on this
      project in the previous month. Enter the amount of dollars that can be
      deducted from this month's budget so that the hours achieved looks more
      realistic.

    """
    project = models.ForeignKey(
        Project, verbose_name=('Project'), related_name='project_months')
    month = models.ForeignKey(Month, verbose_name=('Month'))
    budget = models.FloatField(verbose_name=_('Budget'))
    rate = models.FloatField(verbose_name=_('Rate'))
    overhead_previous_month = models.FloatField(
        verbose_name=_('Overhead previous month'), blank=True, null=True)

    objects = ProjectMonthManager()

    class Meta:
        ordering = ['-month', 'project']

    def __str__(self):
        return '{0} - {1}'.format(self.project, self.month)

    def get_budget_hours(self):
        """Returns the budget hours based on the budget and the rate."""
        overhead = self.overhead_previous_month
        if overhead is None:
            overhead = 0
        return (self.budget - overhead) / self.rate

    def get_daily_hours(self):
        """Returns the daily hours needed in order to use up the budget."""
        budget_hours = self.get_budget_hours()
        return budget_hours / self.month.get_work_days()


@python_2_unicode_compatible
class EmployeeProjectMonth(models.Model):
    project_month = models.ForeignKey(
        ProjectMonth, verbose_name=_('ProjectMonth'),
        related_name='employee_project_months')
    employee = models.ForeignKey(
        Employee, verbose_name=_('Employee'),
        related_name='employee_project_months')
    responsibility = models.PositiveIntegerField(
        verbose_name=_('Responsibility'),
        help_text=_('Positive integer (1-100%)'))

    class Meta:
        ordering = ['project_month__project', 'employee']

    def __str__(self):
        return '{0} - {1}'.format(self.project_month, self.employee)
